Prepare docs for 10 release (#415)
preactjs/preact-www: content/en/guide/v10/whats-new.md

---
name: What's new in Preact X
description: 'New features and changes in Preact X'
---

# What's new in Preact X

> Preact X is a huge step forward from Preact 8.x. We've rethought every bit and byte of our code and added a plethora of major features in the process. Same goes to for compatibility enhancements to support more third-party libraries.

Reviewer: "Same goes to for" -> "Same goes for"?

[Lifecycle docs →](/guide/v10/components#componentdidcatch)

```jsx
class Catcher extends Component {
  state = { errored: false }

  componentDidCatch(error) {
    this.setState({ errored: true });
  }

  render(props, state) {
    if (state.errored) {
      return <p>Something went badly wrong</p>;
    }
    return props.children;
  }
}
```

## Hooks

`Hooks` are a new way to make sharing logic easier between components. They represent an alternative to the existing class-based component API. In Preact they live inside an addon which can be imported via `preact/hooks`

[Hooks Docs →](/guide/v10/hooks)

```jsx
function Counter() {
  const [value, setValue] = useState(0);
  const increment = useCallback(() => setValue(value + 1), [value]);

  return (
    <div>
      Counter: {value}
      <button onClick={increment}>Increment</button>
    </div>
  );
}
```

## createContext

> The `createContext`-API is a true successor for `getChildContext()`. Whereas `getChildContext` is fine when you're absolutely sure to never change a value, it falls apart as soon as a component in-between the provider and consumer blocks an update via `shouldComponentUpdate` when it returns `false`. With the new context API this problem is now a thing of the past. It is a true pub/sub solution to deliver updates deep down the tree.

Author: It's better than the current one, so I'll commit it :)

[createContext Docs →](/guide/v10/context#createcontext)

```jsx
const Theme = createContext('light');

function ThemedButton(props) {
  return (
    <Theme.Consumer>
      {theme => <div>Active theme: {theme}</div>}
    </Theme.Consumer>
  );
}

function App() {
  return (
    <Theme.Provider value="dark">
      <SomeComponent>
        <ThemedButton />
      </SomeComponent>
    <Theme.Provider />
  );
}
```

## CSS Custom Properties

> Sometimes it's the little things that make a huge difference. With the recent advancements in CSS you can leverage variables for styling:

Reviewer 2: ```suggestion
Sometimes it's the little things that make a huge difference. With the recent advancements in CSS you can leverage [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/--*) for styling:
```

```jsx
function Foo(props) {

>   return <div> style={{ --theme-color: "blue" }}>{props.children}</div>;

Reviewer 2: ```suggestion
  return <div> style={{ '--theme-color': 'blue' }}>{props.children}</div>;
```

}
```

## Compat lives in core

> Although we were always keen on adding new features and pushing Preact forward, the `preact-compat` package didn't receive as much love. Up until now it has lived in a separate repository making it harder to introduce breaking changes.

Reviewer 2: ```suggestion
Although we were always keen on adding new features and pushing Preact forward, the `preact-compat` package didn't receive as much love. Up until now it has lived in a separate repository making it harder to coordinate large changes spanning Preact and the compatibility layer. By moving compat into the same package as Preact itself, there's nothing extra to install in order to use libraries from the React ecosystem.
```

> [Compat](/guide/v10/differences-to-react#features-exclusive-to-preactcompat) has learned several new tricks such as `forwardRef`, `memo` and countless compatibility improvements.

Reviewer 2: ```suggestion
The compatibility layer is now called [preact/compat](/guide/v10/differences-to-react#features-exclusive-to-preactcompat), and has learned several new tricks such as `forwardRef`, `memo` and countless compatibility improvements.
```
